src/kartograf/filters/ring_changes.py

```diff
@@ -162,3 +163,55 @@ def get_atom_ring_hybridization_map(rdmol: Chem.Mol) -> dict[int, bool]:
             filtered_mapping[ai] = aj
 
     return filtered_mapping
+
+def filter_fused_ring_changes(molA: Chem.Mol, molB: Chem.Mol, mapping: ATOM_MAPPING) -> ATOM_MAPPING:
+    """
+    Remove cases where a fused ring is partially mapped and could be considered broken
+    see <https://github.com/OpenFreeEnergy/kartograf/pull/56> for more details.
+    """
+    proposed_mapping = {**mapping}
+
+    for mol in [molA, molB]:  # loop over A->B and B->A directions
+        ri = (
+            mol.GetRingInfo().AtomRings()
+        )  # gives list of tuple of atom indices
+        # for each ring, are we fully present?
+        ringok: dict[frozenset[int], bool] = {}
+        # for each atom, which (maybe unmapped) rings is it present in?
+        atom2ring: dict[int, list[frozenset[int]]] = defaultdict(list)
+        for ring in ri:
+            ringset = frozenset(ring)
+
+            ringok[ringset] = all(atom in proposed_mapping for atom in ring)
+            for atom in ring:
+                atom2ring[atom].append(ringset)
+
+        filtered_mapping = {}
+        for i, j in proposed_mapping.items():
+            # if not in any rings, we're ok
+            if i not in atom2ring:
+                filtered_mapping[i] = j
+                continue
+
+            # all rings the atom is present must be in the mapping
+            # this will result in broken fused ring mappings
+            if all(ringok[r] for r in atom2ring[i]):
+                filtered_mapping[i] = j
+
+        # remove partial rings from the mapping which are fused
+        to_remove = []
+        for ring in ri:
+            # if every atom in each ring is not mapped remove the entire ring
+            if not all(r in filtered_mapping for r in ring):
+                to_remove.extend(ring)
+        for i in set(to_remove):
+            try:
+                # some of the ring atoms were not present as we are removing partial rings
+                del filtered_mapping[i]
+            except KeyError:
+                continue
+
+        # reverse the mapping to check B->A (then reverse again)
+        proposed_mapping = {v: k for k, v in filtered_mapping.items()}
```

Comment:
I will say this makes me a bit nervous that someone's going to mess with the loop and get this going the wrong way 😅 

Maybe at the very least add a comment on line 173 that tells folks not to reverse the order?

Reply:
Yeah its a bit risky but the tests should catch this!